Add biorthogonal wavelets

niinimaki opened the pull request:
Please review the implementations.
Especially the importing of `jos_wavelets`.

adler-j commented:
I would personally put all code associated with `jos_wavelets` into his library and add `odl` as a dependency to it, as it is now, I doubt it will have many external users until it stabilizes. Adding very young codes straight into ODL will cause us some headaches in the long run since we then risk becoming the point of support of these.

Much of the code could be moved "as is" to that library, so I'll add some comments on this.

adler-j commented on odl/trafos/wavelet.py:
```diff
@@ -653,22 +704,29 @@ def _call(self, x):
+                                        wbasis=self.wbasis)
+            coeff = Op(x)
+            return self.range.element(coeff)
+        elif isinstance(self.wbasis, pywt.Wavelet):
```
this could have been written `else:`

adler-j commented on odl/trafos/wavelet.py:
```diff
@@ -653,22 +704,29 @@ def _call(self, x):
+                coeff_list = pywt.wavedec2(x, self.wbasis, self.mode,
+                                           self.nscales)
+                coeff_arr = pywt_coeff_to_array(coeff_list, self.size_list)
+                return self.range.element(coeff_arr)
```
This is automatically done by odl, simply do:

```python
coeff_list = pywt.wavedec2(x, self.wbasis, self.mode,
                           self.nscales)
return pywt_coeff_to_array(coeff_list, self.size_list)
```

adler-j commented on odl/trafos/wavelet.py:
```diff
@@ -653,22 +704,29 @@ def _call(self, x):
+            Op = BiorthWaveletTransform(dom=self.domain, nscales=self.nscales,
+                                        wbasis=self.wbasis)
+            coeff = Op(x)
+            return self.range.element(coeff)
```
What is the range of `Op`, does this need to be called?

niinimaki replied:
The range of `BiorthWaveletTransform` doesn't need to be called, the range is determined inside the class. Now the user only needs to know the "image domain" and thus the calling of  `BiorthWaveletTransform` is uniform with  calling of `WaveletTransform`.

adler-j commented on odl/trafos/wavelet.py:
```diff
@@ -599,45 +630,65 @@ def __init__(self, dom, nscales, wbasis, mode):
         """Whether or not the wavelet basis is bi-orthogonal."""
-        return self.wbasis.biorthogonal
+        if isinstance(self.wbasis, pywt.Wavelet):
+            return self.wbasis.biorthogonal
```
else?

adler-j commented on odl/trafos/wavelet.py:
```diff
@@ -25,17 +25,25 @@
 try:
     import pywt
     PYWAVELETS_AVAILABLE = True
+    import jos_wavelets as bwt
```
You need two try catches to do things like this:

```python
try:
    import lib1
    LIB1_AVAILABLE = True    
except ImportError:
    LIB1_AVAILABLE = False
    
try:
    import lib2
    LIB2_AVAILABLE = True    
except ImportError:
    LIB2_AVAILABLE = False
```

adler-j commented on odl/trafos/wavelet.py:
```diff
@@ -532,26 +541,37 @@ def __init__(self, dom, nscales, wbasis, mode):
             Possible wavelet families are:
 
-            Haar (``haar``)
+            Haar (``'haar'``)
```
These changes to doc are all good

adler-j commented on odl/trafos/wavelet.py:
```diff
@@ -676,18 +734,28 @@ def adjoint(self):
+            return InverseBiorthWaveletTransform(ran=self.domain,
+                                                 nscales=self.nscales,
+                                                 wbasis=self.wbasis)
+        elif isinstance(self.wbasis, pywt.Wavelet):
```
Just else

adler-j commented on odl/trafos/wavelet.py:
```diff
@@ -750,11 +828,26 @@ def __init__(self, ran, nscales, wbasis, mode):
+            try:
+                wbasis = wbasis + ''
+                if wbasis.startswith('JOS'):
+                    self.wbasis = wbasis
```
Here `self.wbasis` can be either a `str` or a `pywt.Wavelet` such type mixings are in general not good since any consumer of this property would need to check which one it is at all times.

niinimaki replied:
I suppose you refer mainly to ` wbasis = wbasis + ''`
This was a fix suggested by @kohr-h, previously I had
 
 `if isinstance(wbasis, pywt.Wavelet):`
            `self.wbasis = wbasis`
        `elif isinstance(wbasis, str):`
            `if wbasis[0:6] == 'biorth':`
                `self.wbasis = wbasis`
            `else:`
                `self.wbasis = pywt.Wavelet(wbasis)`

but this did not work (and would have certainly needed some improve). 

As you said `self.wbasis` can be either a `str` or a `pywt.Wavelet`. When it is a `str` one needs to distinguish if it refers to the wavelet families within pyWavelets (quite a many combinations with all possible filters in all wavelet families)  of to the Jan-Olov's wavelets. If it refers to Jan-Olov's then it will further be passed as a `str` else transformed to `pywt.Wavelet`. As I said my previous attempt passed all `str` to `BiorthWaveletTranform` even when they referred to some of pyWavelets wavelet families...

Would appreciate a comment how to better formulate this.

kohr-h replied:
The simplest possibility, and I guess that's what @adler-j means, is to always expect a string and only differentiate between `wbasis.startswith('jos')` (lowercase preferred). For the pywavelets strings, you would use the string to initialize a `Wavelet` object, otherwise you call the corresponding function in Jan-Olov's library.

In the long run, I suggest to use a flag like `'backend'` with choices `pywavelets` or `jos` and separate the wavelet families according to that. Then add a property `backend` to the classes, that makes checking for implementation easy. You can also split the `_call` code accordingly, i.e. make a `_call_pyfftw` and a `_call_jos` and just call those from the `_call` depending on backend.

adler-j replied:
I.m.o calls to two different libraries should not be in the same file, this makes the imports etc confusing. I think the way it has been done with `XrayTransform` is ok, make a general class and make it call an `...impl` in some other file which contains all the dirty inner workings.

kohr-h replied:
Right, we could actually just clone that pattern with a `backends` directory and push the `size_list` and stuff to the `pywavelets` part and `jos` to  another file there. And then we would just need three operators here, `WaveletTransform`, `WaveletTransformAdjoint` and `WaveletTransformInverse`. Regarding names, we have somehow settled for this pattern since it will put the names together in auto-completion lists. Sorry @niinimaki, I know I said differently a while ago. I do the restructuring.

adler-j commented on odl/trafos/wavelet.py:
```diff
@@ -763,44 +856,66 @@ def __init__(self, ran, nscales, wbasis, mode):
         """Whether or not the wavelet basis is bi-orthogonal."""
-        return self.wbasis.biorthogonal
+        if isinstance(self.wbasis, pywt.Wavelet):
+            return self.wbasis.biorthogonal
```
else?

adler-j commented on odl/trafos/wavelet.py:
```diff
@@ -834,14 +957,511 @@ def adjoint(self):
+                                    wbasis=self.wbasis, mode=self.mode)
+
+
+class BiorthWaveletTransform(Operator):
```
Why would we have a separate `BiorthWaveletTransform` if `WaveletTransform` can already be biorthonormal? It even has a `is_biorthogonal` property. This needs to be documented.

niinimaki replied:
The separate `BiorthWaveletTransform` is due to the Jan-Olov's wavelet library.
The biorthogonal wavelets that are in pyWavelets have even filter lengths, Jan-Olov's wavelets have odd filter length. Also for Jan-Olov's wavelets there is own classes for the adjoint transform and the inverse adjoint transform, these we don't have for biorthogonal pyWavelets (with orthogonal wavelets they are not necessary). _Maybe_ Jan-Olov will write adjoint and inverse adjoint codes for even length filters at some point, but that remains to be seen...

I'll add some comment to the code related to this

kohr-h replied:
How about adding this to the standard `WaveletTransform` and just raise an error if the filter length is not supported? The init patterns are quite similar, so no worries there. You only have to make `mode` optional (which would be good anyway), expect `wbasis` to be a string, and we're all good I guess.

adler-j replied:
I think the main line of thought here still should be to move all of this JOS things outside of ODL and into a fully separate operator, then it could be called `BiorthWaveletTransform` without problems.

If we wanted to, we could make an abstract "WaveletTransform" class that both of these inherit from (we could even overload its `__new__` to return a proper one depending on inputs).

kohr-h replied:
> to return a proper one depending on inputs

What does that mean?

adler-j replied:
Something like:

```python
class WaveletTransform(object):
    def __new__(cls, arg=None):
        if arg == 'jos':
            cls = JOSWaveletTransform
        elif arg == 'pywt':
            cls = pywaveletWaveletTransform
        return object.__new__(cls)
        
    def general_wavelet_method(self):
        pass


class JOSWaveletTransform(WaveletTransform):
    def __repr__(self):
        return 'JosWaveletTransform()'
        
    def JOS_specific_method(self):
        pass


class pywaveletWaveletTransform(WaveletTransform):
    def __repr__(self):
        return 'pywaveletWaveletTransform()'
```        

with calls working like this

```python
>>> print(WaveletTransform('jos'))
JosWaveletTransform()
>>> print(JOSWaveletTransform())
JosWaveletTransform()
>>> print(WaveletTransform('pywt'))
pywaveletWaveletTransform()
>>> print(pywaveletWaveletTransform())
pywaveletWaveletTransform()
```

kohr-h replied:
Oh, you're dusting off the old idea with the top-level factory class. Yeah, perhaps.

Side note: I know that `__repr__` is just an example, but one should clearly use `self.__class__.__name__` there.

adler-j commented on odl/trafos/wavelet.py:
```diff
@@ -834,14 +957,511 @@ def adjoint(self):
+
+        ndim = dom.ndim
+        if ndim == 1:
+            max_level = np.log2(dom.grid.shape[0])
```
log2 returns a float, how is rounding done here?

niinimaki replied:
After the `if` statements the rounding is done by

`max_level = np.ceil(max_level)`

kohr-h replied:
You can simply do `max_level = int(np.ceil(np.log2(np.max(dom.shape))))`. The dimension check can happen independently.

adler-j replied:
:+1:

adler-j commented on odl/trafos/wavelet.py:
```diff
@@ -834,14 +957,511 @@ def adjoint(self):
+                             'got {}.'.format(max_level, self.nscales))
+            #raiseETC ('Setting {} to {}' .format(self.nscales, max_level))
+
+        filterlength = int(wbasis[9])
```
`filterlength = int(wbasis[-1])` would perhaps be more clear

adler-j commented on odl/trafos/wavelet.py:
```diff
@@ -834,14 +957,511 @@ def adjoint(self):
+            raise NotImplementedError('Filterlength {} not 1, 3, 5, 7 or 9'
+                                      ''.format(filterlength))
+
+        ran_size = dom.size
```
redundant variable

adler-j commented on odl/trafos/wavelet.py:
```diff
@@ -834,14 +957,511 @@ def adjoint(self):
+            The length of the array depends on the size of input image to
+            be transformed and on the chosen wavelet basis.
+        """
+        filterlength = int(self.wbasis[9])
```
If this is used repeatedly perhaps we can save it

adler-j commented on odl/trafos/wavelet.py:
```diff
@@ -834,14 +957,511 @@ def adjoint(self):
+
+        wbasis : `str`
+            (``'JOSbiorthN'``) J.-O. Stomberg's biorthogonal wavelets
+            where ``N`` refers to the filter length.
```
Why does `nscales` get its own parameter but `N` does not?

niinimaki replied:
This is again due to making the calling of `WaveletTransform` and `BiorthWaveletTransform` uniform. Also `nscales` refers to the number of scaling levels that can be chosen independently from the choice of wavelets (i.e. almost independently, up to the maximum useful level that is determined from the size of the image and from the chosen wavelet basis). `N` refers to the filter length that are implemented in Jan-Olov's wavelet library and the values are fixed to `1, 3, 5, 7` or `9`. `nscales` is used for both `WaveletTransform` and `BiorthWaveletTransform` but `N` would only be needed for  `BiorthWaveletTransform`

adler-j commented on odl/trafos/wavelet.py:
```diff
@@ -834,14 +957,511 @@ def adjoint(self):
+            #A copy of coeff is unavoidable since bwt writes over
+            #the given input
+            coeff_in = coeff.copy()
+            coeff_in = coeff_in.asarray()
```
simply do this on one line:

``` 
coeff_in = coeff.copy().asarray()
```

adler-j commented on odl/trafos/wavelet.py:
```diff
@@ -834,14 +957,511 @@ def adjoint(self):
+            #the given input
+            coeff_in = coeff.copy()
+            coeff_in = coeff_in.asarray()
+            x = np.asarray(self.range.element())
```
Decide on either `vector.asarray()` or `np.asarray(vector)`, but try to be consistent.

adler-j replied:
`x` does not realy convey what it is intended to do

adler-j commented on odl/trafos/wavelet.py:
```diff
@@ -834,14 +957,511 @@ def adjoint(self):
+            raise ValueError('domain Lp exponent is {} instead of 2.0.'
+                             ''.format(ran.exponent))
+
+        ndim = ran.ndim
```
This piece of code (following 10 lines) is written many times. Perhaps make a utility?

adler-j commented on odl/trafos/wavelet.py:
```diff
@@ -834,14 +957,511 @@ def adjoint(self):
+                                      wbasis=self.wbasis)
+
+
+class InverseAdjBiorthWaveletTransform(Operator):
```
These four classes are all terribly similar, is there no chance to change them to one class?

niinimaki commented:
> I would personally put all code associated with jos_wavelets into his library and add odl as a dependency to it, as it is now, I doubt it will have many external users until it stabilizes. Adding very young codes straight into ODL will cause us some headaches in the long run since we then risk becoming the point of support of these.Much of the code could be moved "as is" to that library, so I'll add some comments on this.

It is OK for me to put these codes into `jos_wavelets` library, if it is preferred. 
So, how do we proceed then?
 But anyway I'll first do the modifications according to the review.

> Also, the name of the commit is somewhat non-descriptive.

Sorry about that, will be more descriptive in the future.

kohr-h commented on odl/trafos/wavelet.py:
```diff
@@ -834,14 +957,511 @@ def adjoint(self):
+        nscales : `int`
+            Number of scaling levels in the wavelet transform.
+        wbasis : `str`
+            (``'JOSbiorthN'``) J.O. Stomberg's biorthogonal wavelets
```
"r" in Strömberg missing.

kohr-h commented on odl/trafos/wavelet.py:
```diff
@@ -834,14 +957,511 @@ def adjoint(self):
+            max_level = np.log2(max(dom.grid.shape[0], dom.grid.shape[1],
+                                dom.grid.shape[2]))
+        else:
+            raise NotImplementedError('ndim {} not 1, 2 or 3'
```
Since the user never specified something like `ndim`, this error message will look cryptic. Better say something like 
`'dimension of input is {}, supported are 1, 2 and 3.'.format(dom.ndim)`. `len(dom.ndim)` will raise an error BTW.

niinimaki commented:
The following modifications have been made:

- Replaces `elif`s with `else` where applicable and added missing/clarifying `else` statements where necessary.

- removed redundant variables (`ndim` ->`dom.ndim` or `ran.ndim` accordingly, removed redundant `ran_size` `dom_size` where applicable, new deffinition of `max_level` related jos_wavelets removed a lot of redundancies)

- improved error messaged according to @kohr-h suggestion

- changes to lowercase letters `JOS` -> `jos` in code and in the test (and corrected the spelling error/erros)

- `filterlength = int(wbasis[9])` -> `filterlength = int(wbasis[-1])` and saved the variable in `BiorthWaveletTransform` classes

- `try` - `import` - `except` also modified in `wavelet.py`

Test seem to fail due to ` NameError: global name 'pywt' is not defined`...

I suppose that the next (big) thing would then be the restructuring @kohr-h mentioned...

kohr-h commented:
> The following modifications have been made:

Being picky now: please don't use `ENH` in all commit messages. That one is not intended for general "improvement" of code, which could be anything, but rather real improvements, e.g. speed optimizations, feature additions etc.
What you did was mostly refactoring, removal of redundant code, logic simplifications (these could possibly be ENH), spelling fixes, which are MAINT. If you fix a bug, use BUG. If you add a test, use TST. If you modify/add documentation, use DOC.

> Test seem to fail due to NameError: global name 'pywt' is not defined...

A bunch of tests fail since `bwt` is not skipped if not present. You need to add a skip decorator similar to the `skip_if_no_pywavelets`. If there are also tests with `pywavelets` failing, check if all those are decorated properly.

> I suppose that the next (big) thing would then be the restructuring @kohr-h mentioned...

I'll look into it.

niinimaki commented:
> What you did was mostly refactoring, removal of redundant code, logic simplifications (these could possibly be ENH), spelling fixes, which are MAINT. If you fix a bug, use BUG. If you add a test, use TST. If you modify/add documentation, use DOC.

Thanks for the reminder on proper commits :)

> You need to add a skip decorator similar to the skip_if_no_pywavelets.

Besides addind `@skip_if_no...` in the wavelet_test.py I then need to add something like

`    skip_if_no_joswavelets = pytest.mark.skipif(`
`        "not odl.trafos.wavelet.JOSWAVELETS_AVAILABLE",`
`        reason='jos_wavelet not available'`
    )

in odl.utils.testutils, right? Are there other places where I should add lines related to skipping if `bwt` not available??

kohr-h commented:
> Besides addind @skip_if_no... in the wavelet_test.py I then need to add something like
> 
> skip_if_no_joswavelets = pytest.mark.skipif(
> "not odl.trafos.wavelet.JOSWAVELETS_AVAILABLE",
> reason='jos_wavelet not available'
> )
> 
> in odl.utils.testutils, right? 

Yes. The tests and this decorator will later go to JO's repo, together with the rest.

> Are there other places where I should add lines related to skipping if bwt not available??

Only in the tests that are using it.

niinimaki commented:
Actually I think that we do not need

`    @property`
`    def is_biorthogonal(self):`
`...`

since it is not used afterward unlike `is_orthogonal`
This could partly reduce the unclearness of having biorthogonal wavelets in pyWavelet but still a separate library for (another type of) biorthogonal wavelets...

Opinions?

adler-j commented:
In general we prefer a `is_...` method over adding a new class (the removal of `LinearOperator` ages ago is an example of this). The reason is that if you make a new class, users have to use it and they have to do more work do distinguish cases:

```python
if operator.is_linear:
    pass
```

vs

```python
if isinstance(operator, odl.LinearOperator):
    pass
```

And you also get some real issues for corner cases where it is conditionally linear:

```python
class AdditionOperator(odl.Operator)
    """ Calculates ax + b, is linear of b=0"""
    def __init__(self, a, b):
        super().__init__(RealNumbers(), RealNumbers(), linear=(b==0))
    def _call(x):
        return a * x + b 
```

vs:

```python
class AdditionOperator(odl.Operator)
    """ Calculates ax + b, is linear of b=0"""
    def __new__(cls, a, b):
        # This is one way to implement it, the other would be to create two classes and a factory
        if b==0:
            # Create a subclass that is linear
            return type('AdditionOperatorLinear', (AdditionOperator, odl.LinearOperator), {})(a, b)
        else:
            return odl.Operator.__new__(cls, a, b)
    def __init__(self, a, b):
        super().__init__(RealNumbers(), RealNumbers())
    def _call(x):
        return a * x + b 
```

I suppose many of these issues could pop up for the biorthonromal case aswell, someone builds a function that takes an arbitrary `WaveletTransform` and returns something, then he would need to make two classes, one for `Biorthonormal` and one for the other.

niinimaki commented:
The following modifications have been made:

- added skips for test if `jos_wavelet` not available

- make `mode` to have a default value in `WaveletTransform` and `InverseWaveletTransform`

- Added more comments about Jan-Olov's wavelets into the docstrings of biorthogonal wavelet classes

kohr-h commented:
Okay, I'll see what I can do regarding the code splitting and refactoring.

adler-j commented:
So what's happening here?

kohr-h commented:
Not so much lately. Code needs to be moved from one repo to the other.

adler-j commented:
I guess we close this? Any chance this comes in?

kohr-h commented:
Well I did some of the work here but honestly I don't quite see the point of putting more time into it. The least we can do is to move the corresponding code to the other repo, albeit in non-working state.

adler-j commented:
If it can be moved "cleanly" then sure go ahead.

adler-j commented:
Did you make a new repo somewhere? Then link here if someone manages to google their way here.

adler-j commented:
https://github.com/odlgroup/jos_wavelets

kohr-h commented:
Oh, it's been around for a while. I just pushed over the stuff from here now.
